crypto/threads_none.c

```diff
@@ -0,0 +1,138 @@
+ *    for use in the OpenSSL Toolkit. (http://www.openssl.org/)"
+ *
+ * 4. The names "OpenSSL Toolkit" and "OpenSSL Project" must not be used to
+ *    endorse or promote products derived from this software without
+ *    prior written permission. For written permission, please contact
+ *    openssl-core@openssl.org.
+ *
+ * 5. Products derived from this software may not be called "OpenSSL"
+ *    nor may "OpenSSL" appear in their names without prior written
+ *    permission of the OpenSSL Project.
+ *
+ * 6. Redistributions of any form whatsoever must retain the following
+ *    acknowledgment:
+ *    "This product includes software developed by the OpenSSL Project
+ *    for use in the OpenSSL Toolkit (http://www.openssl.org/)"
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE OpenSSL PROJECT ``AS IS'' AND ANY
+ * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
+ * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ * ====================================================================
+ */
+
+#include <openssl/crypto.h>
+
+#if !defined(OPENSSL_THREADS) || defined(CRYPTO_TDEBUG)
+
+int CRYPTO_MUTEX_init(CRYPTO_MUTEX *lock)
+{
+    int *l = lock;
+    *l = 1;
+
+    return 1;
+}
+
+int CRYPTO_MUTEX_lock_read(CRYPTO_MUTEX *lock)
+{
+    OPENSSL_assert(*(int *) lock == 1);
+    return 1;
+}
+
+int CRYPTO_MUTEX_lock_write(CRYPTO_MUTEX *lock)
+{
+    OPENSSL_assert(*(int *) lock == 1);
+    return 1;
+}
+
+int CRYPTO_MUTEX_unlock(CRYPTO_MUTEX *lock)
+{
+    OPENSSL_assert(*(int *) lock == 1);
+    return 1;
+}
+
+void CRYPTO_MUTEX_cleanup(CRYPTO_MUTEX *lock) { }
+
+void CRYPTO_ONCE_run(CRYPTO_ONCE *once, void (*init)(void))
+{
+    if (*once)
+        return;
+
+    init();
+    *once = 1;
+}
+
+#define OPENSSL_CRYPTO_THREAD_LOCAL_KEY_MAX 256
+
+static void *thread_local_storage[OPENSSL_CRYPTO_THREAD_LOCAL_KEY_MAX];
+static void *thread_local_cleanup[OPENSSL_CRYPTO_THREAD_LOCAL_KEY_MAX];
+
+int CRYPTO_THREAD_LOCAL_init(CRYPTO_THREAD_LOCAL *key, void (*cleanup)(void *))
+{
+    static int thread_local_key = 0;
+
+    if (thread_local_key >= OPENSSL_CRYPTO_THREAD_LOCAL_KEY_MAX)
+        return 0;
+
+    *key = thread_local_key++;
+
+    thread_local_cleanup[*key] = cleanup;
+
+    return 1;
+}
+
+void *CRYPTO_THREAD_LOCAL_get(CRYPTO_THREAD_LOCAL *key)
+{
+    if (*key >= OPENSSL_CRYPTO_THREAD_LOCAL_KEY_MAX)
```

Comment:
Need to check that *key is nonnegative, too, at which point a local variable to hold *key may make sense.
Or just change CRYPTO_THREAD_LOCAL to an unsigned type ;)